0.preprocess-sites/8.image-and-segmentation-qc.py

```diff
@@ -0,0 +1,406 @@
+    build_seq = list(
+        zip(
+            ([i] * (sites_per_image_grid_side + 1)),
+            reversed(range(1, (sites_per_image_grid_side + 1))),
+        )
+    )
+    final_order += build_seq
+
+# Uses sites_list in case there are fewer analyzed sites than acquired sites
+sites_list = [*range(1, (sites_per_image_grid_side * sites_per_image_grid_side) + 1)]
+loc_df = (
+    pd.DataFrame(final_order)
+    .rename(columns={0: "x_loc", 1: "y_loc"})
+    .assign(Site=sites_list)
+)
+
+# Create total_cell_count
+cell_count_bysite_df = (
+    cell_count_df.groupby("site_full")["cell_count"]
+    .sum()
+    .reset_index()
+    .rename(columns={"cell_count": "total_cell_count"})
+)
+
+# Add total_cell_count to cell_count_df and add in x, y coordinates for plotting
+cell_count_df = cell_count_df.merge(cell_count_bysite_df, on="site_full").merge(
+    loc_df, on="Site"
+)
+
+# Plot total number of cells per well
+cell_count_totalcells_df = (
+    cell_count_df.groupby(["x_loc", "y_loc", "Well", "Site"])["total_cell_count"]
+    .mean()
+    .reset_index()
+)
+os.makedirs(figures_output, exist_ok=True)
+
+by_well_gg = (
+    gg.ggplot(cell_count_totalcells_df, gg.aes(x="x_loc", y="y_loc"))
+    + gg.geom_point(gg.aes(fill="total_cell_count"), size=10)
+    + gg.geom_text(gg.aes(label="Site"), color="lightgrey")
+    + gg.facet_wrap("~Well")
+    + gg.coord_fixed()
+    + gg.theme_bw()
+    + gg.theme(
+        axis_text=gg.element_blank(),
+        axis_title=gg.element_blank(),
+        strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
+    )
+    + gg.scale_fill_cmap(name="magma")
+)
+output_file = pathlib.Path(figures_output, "plate_layout_cells_count_per_well.png")
+by_well_gg.save(output_file, dpi=300, verbose=False)
+by_well_gg
+
+# Plot cell category ratios per well
+ratio_df = pd.pivot_table(
+    cell_count_df,
+    values="cell_count",
+    index=["site_full", "Plate", "Well", "Site", "x_loc", "y_loc"],
+    columns=["Cell_Quality"],
+)
+ratio_df = ratio_df.assign(
+    Sum=ratio_df.sum(axis=1), Pass_Filter=ratio_df[cell_filter].sum(axis=1)
+)
+fail_filter = [cat for cat in cell_category_order if cat not in cell_filter]
+fail_filter_noempty = [
+    cat for cat in cell_category_order if cat not in cell_filter if cat != "Empty"
+]
+ratio_df = ratio_df.assign(
+    Fail_Filter=ratio_df[fail_filter].sum(axis=1),
+    Fail_Filter_noempty=ratio_df[fail_filter_noempty].sum(axis=1),
+)
+ratio_df = ratio_df.assign(
+    Pass_Fail_withempty=ratio_df["Pass_Filter"] / ratio_df["Fail_Filter"],
+    Pass_Fail_0empty=ratio_df["Pass_Filter"] / ratio_df["Fail_Filter_noempty"],
+    Percent_Empty=ratio_df["Empty"] / ratio_df["Sum"],
+)
+ratio_df = (
+    ratio_df.drop(
+        cell_category_order
+        + ["Sum", "Pass_Filter", "Fail_Filter", "Fail_Filter_noempty"],
+        1,
+    )
+    .stack()
+    .to_frame()
+    .reset_index()
+    .rename(columns={0: "value"})
+)
+ratio_gg = (
```

Comment:
```python
ratio_gg = (
    gg.ggplot(ratio_df, gg.aes(x="x_loc", y="y_loc"))
    + gg.geom_point(gg.aes(fill="Ratio"), size=5)
    + gg.geom_text(gg.aes(label="Site"), size=4, color="lightgrey")
    + gg.facet_grid("Cell_Quality~Well", scales="free_y")
    + gg.coord_fixed()
    + gg.theme_bw()
    + gg.ggtitle("Quality Ratio")
    + gg.coord_fixed()
    + gg.theme(
        axis_text=gg.element_blank(),
        axis_title=gg.element_blank(),
        strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
    )
    + gg.scale_fill_cmap(name="magma")
)

output_file = pathlib.Path("plate_layout_ratios_per_well.png")
ratio_gg.save(output_file, dpi=300, width=5, height=6, verbose=False)
```

One thing I didn't do here was fix the hardcoded `width`. I recommend using `num_wells + 1`. I didn't do it here, b/c I did notice that some of the plots had different wells. Maybe this is just b/c I'm working with test data, but sometimes the well is A1 and A2, while other times it was B1 and B2.

Reply:
I've got one more update I'd like for this plot that I can't figure out. I'd like for labeller to use this dictionary to update the row name.  `{'Pass_Fail_withempty': 'Pass/Fail \n with empty', 'Percent_Empty':'Percent Empty' 'Pass/Fail \n without empty': 'Pass/Fail \n without empty'}`

Reply:
I don't know of a way to manipulate the "strip text" except outside the plot. This works:

```python
quality_recode = {
    "Pass_Fail_withempty": "Pass/Fail (with empty)",
    "Pass_Fail_0empty": "Pass/Fail (without empty)",
    "Percent_Empty":"Percent Empty",
}
ratio_df = ratio_df.assign(cell_quality_recode=ratio_df.Cell_Quality.replace(quality_recode))

ratio_gg = (
    gg.ggplot(ratio_df, gg.aes(x="x_loc", y="y_loc"))
    + gg.geom_point(gg.aes(fill="Ratio"), size=5)
    + gg.geom_text(gg.aes(label="Site"), size=4, color="lightgrey")
    + gg.facet_grid("cell_quality_recode~Well", scales="free_y")
    + gg.coord_fixed()
    + gg.theme_bw()
    + gg.ggtitle("Quality Ratio")
    + gg.coord_fixed()
    + gg.theme(
        axis_text=gg.element_blank(),
        axis_title=gg.element_blank(),
        strip_background=gg.element_rect(colour="black", fill="#fdfff4"),
        strip_text_y=gg.element_text(size=7),
    )
    + gg.scale_fill_cmap(name="magma")
)
```

Reply:
oh, also note that I removed the new line character in favor of parentheses. Strip text kinda looks weird on two lines. Go with what you prefer though.